service/dap/server.go

```diff
@@ -130,6 +130,19 @@ type Server struct {
 	// sendingMu synchronizes writing to net.Conn
 	// to ensure that messages do not get interleaved
 	sendingMu sync.Mutex
+
+	// running tracks the running state according to the server,
+	// which may not correspond to the actual running state of
+	// the process.
+	running   bool
+	runningMu sync.Mutex
+
+	// resumeMu must be held for a request to resume execution of the
+	// process.
+	// We use this to guarantee that no other request can unexpectedly
+	// change the state to running, when we expect the process to halted.
+	// Do not attempt to lock if already holding sendingMu or runningMu.
```

Comment:
I was using this comment to try to document the right order without having to read the code. Is there a better way to do that?

Reply:
I don't know of any special ways. I think it is just the general principle that you always lock multiple locks in the same fixed order and unlock in reverse order to avoid deadlocks. Anybody using this code would also need to take care with mu lock as well.  Coordinating many locks was one of the reason why there was pushback on having them in earlier PRs.